Document/0x06c-Reverse-Engineering-and-Tampering.md

```diff
@@ -593,15 +593,97 @@ Corellium allows you to launch multiple instances of a device (jailbroken or not
+
+### Unicorn
+
+Unicorn is a lightweight, multi-architecture CPU emulator framework. Unicorn provides us the basic infrastructure needed to execute processor instructions. In this section we will use Unicorn to solve UnCrackable Level 1 iOS App.
+
+To use Unicorn, we need to implement all necessary infrastructure which generally is readily available to us in an Operating System, which includes binary loader, linker and other dependency libraries. Writing a complete binary loader is superfluous for our current goal,  instead we will write our own crude binary loader and execute a small part of the binary in Unicorn to solve the challenge.
+
+While performing manual analysis in "[Reviewing Disassembled Native Code](#reviewing-disassembled-native-code "Reviewing Disassembled Native Code")" section, we determined that the function at offset 0x1000080d4 is responsible for dynamically generating the secret string. If we analyze this function and the subsequent function calls, we will observe that there is no hard dependency on any external library and neither its performing any system calls. All the necessary code is self-contained in the binary and this is a perfect scenario to use a CPU emulator like Unicorn.
+
+To solve the challenge using Unicorn we will broadly perform the following steps:
+
+- Get the ARM64 version of the binary by running `lipo -thin arm64 <app_binary> -output uncrackable.arm64` (ARMv7 can be used as well).
+- Extract the `__text` and `__data` section from the binary.
+- Create and map the memory to be used as stack memory.
+- Create memory and load the `__text` and `__data` section.
+- Execute the binary by providing the start and end address.
+- Finally, dump the return value from the function, which in this case is our secret string.
+
+In the present scenario, apart from the `__text` section (which contains the instructions) we also need to load `__data` section. The program is accessing some hard coded values, which are typically stored in `__data` section of a Mach-O binary. In function at 0x100008158, at address 0x100008198 a value from address 0x10000d8f6 is being accessed. The address 0x10000d8f6 maps to `__data` section of the binary.
+
+To extract the content of `__text` and `__data` section from the Mach-O binary we will use [LIEF](https://lief.quarkslab.com/ "Lief") library. LIEF provides a convenient abstraction to manipulate multiple executable file formats, including Mach-O files. But before this content into the memory, we also need to determine the load address for `__text` and `__data` section. This can be easily determined using any binary analysis tool like Ghidra, Radare2 or IDA Pro.
+
+<img src="Images/Chapters/0x06c/uncrackable_sections.png" width="500px"/>
+
+From the above table, we will use the base address 0x10000432c for `__text` and 0x10000d3e8 for `__data` section to load them at in the memory.
```

Comment:
you can do this in the script so you don't need another extra tool:

```
emu.mem_write(text_section.virtual_address, bytes(text_section.content)) # 0x10000432c
emu.mem_write(data_section.virtual_address, bytes(data_section.content)) # 0x10000d3e8
```

Reply:
We are already using Ghidra heavily in this chapter, I felt it should be fine to just re-reference it. Also, this helps to connect all tools together and how to use them. 

I am still open to your suggestion, if you feel otherwise. Let me know.

Reply:
it's ok, no need to change it :)